src/components/rating/rating.tsx

```diff
@@ -238,27 +241,88 @@ export class Rating
     this.setFocus();
   }
 
-  private updateValue(value: number) {
-    this.value = value;
-    this.calciteRatingChange.emit({ value });
-  }
+  private handleRatingPointerOver = () => {
+    this.isKeyboardInteraction = false;
+  };
 
-  private onKeyboardPressed = (event: KeyboardEvent): void => {
-    if (!this.required && isActivationKey(event.key)) {
-      event.preventDefault();
-      this.updateValue(0);
-    }
+  private handleRatingPointerOut = () => {
+    this.isKeyboardInteraction = true;
+    this.hoverValue = null;
+    this.focusValue = null;
+  };
+
+  private handleRatingFocusLeave = (): void => {
+    this.focusValue = null;
+    this.hasFocus = false;
+    this.hoverValue = null;
   };
 
-  private onFocusChange = (selectedRatingValue: number): void => {
+  private handleRatingFocusIn = (): void => {
+    this.focusValue = this.value > 0 ? this.value : 1;
     this.hasFocus = true;
-    if (!this.required && this.focusValue === selectedRatingValue) {
-      this.updateValue(0);
+    this.hoverValue = this.value > 0 ? this.value : 1;
+  };
+
+  private handleLabelPointerOver = (ev: PointerEvent) => {
+    const newPointerValue = ev.currentTarget["nextElementSibling"]["value"] || this.value || 0;
+    this.hoverValue = newPointerValue;
+    //  when switching from keyboard to mouse this prop change is required to remove the focus styles from the element which are needed when navigating with a keyboard.
+    this.focusValue = null;
+  };
+
+  private handleKeyDown = (ev: KeyboardEvent) => {
+    if (ev.target["nodeName"] === "CALCITE-RATING") {
+      this.isKeyboardInteraction = true;
     } else {
-      this.focusValue = selectedRatingValue;
+      const inputVal = Number(ev.target["value"]);
+      const key = ev.key;
+      const numberKey = Number(key);
+
+      this.emit = true;
+
+      if (isNaN(numberKey)) {
+        switch (key) {
+          case "Enter":
+            this.value = this.value === inputVal ? 0 : inputVal;
+            break;
+          case "ArrowLeft":
+            this.value = inputVal - 1;
+            break;
+          case "ArrowRight":
+            this.value = inputVal + 1;
+            break;
+          default:
+            break;
+        }
+      } else {
+        if (numberKey >= 0 && numberKey <= this.max) {
+          this.value = numberKey;
+        }
+      }
+
+      this.emit = false;
+      this.focusValue = null;
     }
   };
 
+  private handleInputChange = (ev: Event) => {
+    if (this.isKeyboardInteraction === true) {
+      const inputVal = Number(ev.target["value"]);
+      this.focusValue = inputVal;
+      this.hoverValue = inputVal;
+    }
+  };
+
+  private handleLabelPointerDown = (ev: PointerEvent) => {
+    const inputVal = Number(ev.target["parentElement"]["nextElementSibling"]["value"]);
```

Comment:
I did try replacing ev.target with inputFocusRef. this creates a circular dependency error because inputFocusRef relies on this.value to understand it's current context.

Reply:
Lets create another private var for the input element that doesn't rely on the value. Then the dom structure isn't so dependent for this to work.

Reply:
How often is the internal shadow DOM structure going to change?  I could do a complex function where I get the id of the input from the htmlFor and then query the DOM but that's a much more expensive query than simply navigating the DOM object that we already have available. I did simplify the DOM so the object traversal is simplified.

Reply:
> How often is the internal shadow DOM structure going to change

Probably not often but it does happen.

I think its best if we just refactor this to not depend on the value when setting the `inputFocusRef`

```
ref={(el) =>
              (i === 1 || i === this.value) && (this.inputFocusRef = el as HTMLInputElement)
            }
```

Wherever the current code depends on the `inputFocusRef`, we should just add the if statement there.

Reply:
> Wherever the current code depends on the `inputFocusRef`, we should just add the if statement there.

I don't know what you mean here. The inputFocusRef is useful for React users and in the case of focusIn where the inputFocusRef has already been correctly assigned after a repaint. Because inputFocusRef is assigned on repaint and not before, it can not  be used as the basis for new value assignment. While I could refactor this to assign the inputFocusRef as a part of the updateValue method, this would also require a DOM traversal and would be less performant. For these event listeners set on an element other than the input itself which update the component value , a DOM traversal is the most performant solution to getting the correct new value.

Reply:
Sorry, maybe I'm not understanding correctly.

In general, we have been preferably been storing DOM elements after they are created using the `ref` attribute. This is preferred because when the element has been created by the host component, we can just store a reference to it rather than have to query the internal shadowRoot for it (which is less performant) or have to navigate a specific DOM structure that may change or be altered by future feature updates. So it would be preferable to store a reference to the element when its created.

From my understanding of the stencil lifecycle, the elements are created before rendering in the `componentWillRender` lifecycle method and the same element is used on repaint and modified as necessary.

I think the issue here is that the `inputFocusRef` changes depending on the value. We should probably just store each input element so we have a reference to them as needed. 

Maybe `inputFocusRef` could be changed to an array of elements? `inputFocusRef: HTMLInputElement[]`?

Reply:
Thanks for hoping on a call! As we discussed I have refactored the DOM to make DOM traversal simpler. I had added inputFocusRef on the focusIn event handler with a fallback to traverse the shadowDOM if inputFocusRef is unassigned for some edge case reason. For other event handlers simple DOM traversal (without shadowDom query) is the most performant option.